---  
# Question 2

Contributions from Question 2:
- Jason Li: Wrote and tested code

In [2]:
import numpy as np
import matplotlib.pyplot as plt
from scipy.interpolate import lagrange
from scipy.optimize import curve_fit

## 2a) Use the divided-difference formula to construct an interpolating polynomial from the given set of data

In [28]:
x = [0.1, .2, .3, .4]
fx = [ -0.62049958, -0.28398668, 0.00660095, .24842440]
n = 4
table = np.zeros([4, 4])
table[:,0] = fx
for col in range(1,n):
    for row in range(0,n - col):
        table[row][col] = (table[row + 1][col - 1] - table[row][col - 1])/(x[row + col] - x[row])
print(table)

def approx(x_i, matrx, xVal):
    coefs = matrx[0]
    elements = coefs.size
    res = 0
    ## since we have P(x) = a + b(x-.1) + c(x-.1)(x-.2) + d(x-.1)(x-.2)(x-.3)
    ## this is the same as P(x) = a + (x-.1)(b + (x-.2)(c + (x-.3)(d)))
    ## so we can optimize the calculation by starting from the right, adding the ith coefficient, then multiply by (x-(x_(i-1)))
    for i in range(elements - 1, -1, -1):
        res += coefs[i]
        if i > 0:
            res *= (xVal - x[i - 1])
    return res
print("The approximation is: ", approx(x, table, 0.25))




[[-0.62049958  3.365129   -2.2962635  -0.47315167]
 [-0.28398668  2.9058763  -2.438209    0.        ]
 [ 0.00660095  2.4182345   0.          0.        ]
 [ 0.2484244   0.          0.          0.        ]]
-0.13277477437499996


Using Newton's Divided Differences, we get that f(0.25) is approximately -0.13277477437499996

## 2b) add a point, and approximate again